pam_xauth/pam_xauth.8
.\" Copyright 1999 Red Hat, Inc.
.\" Written by Michael K. Johnson <johnsonm@redhat.com>
.TH pam_xauth 8 1999/4/10 "Red Hat" "System Administrator's Manual"
.SH NAME
pam_xauth \- forward xauth keys between users
.SH SYNOPSIS
.B session optional /lib/security/pam_xauth.so \fIarguments\fP
.SH DESCRIPTION
pam_xauth.so is designed to forward xauth keys (sometimes referred
to as "cookies") between users.

Without pam_xauth,
when xauth is enabled and a user uses the \fBsu\fP command to assume
superuser priviledges, that user is not able to run X commands as root
without somehow giving root access to the xauth key used for the
current X session.  pam_xauth solves the problem by forwarding the
key from the user running su (the source user) to the user whose
identity the source user is assuming (the target user) when the session
is created, and destroying the key when the session is torn down.

This means, for example, that when you run \fBsu\fP from an xterm sesssion,
you will be able to run X programs without explicitly dealing with the
xauth command or ~/.Xauthority files.

By default, pam_xauth will only forward keys when the root user is
the target user.  Source users can add the names of other users that
they trust as target users in the file ~/.xauth/export; once they have
done so, even root is not trusted unless it is listed in ~/.xauth/export.
The special username \fB*\fP matches every user, but neither globbing
nor regexp matching is done.  If ~/.xauth/export is an empty file,
pam_xauth will never forward any of the user's xauth keys anywhere.

Similarly, pam_xauth looks at the target user's ~/.xauth/import file
if the target user wants to accept xauth keys from the source user.
Users who do not wish to have their ~/.Xauthority files mangled when
they are target users can create the empty file ~/.xauth/import and
pam_xauth will do nothing to them\(emneither import keys into them
nor remove keys from them.

pam_xauth will only forward keys if xauth can list a key connected
to the $DISPLAY environment variable.

If a source user opens multiple sessions from a single display to a
single target user, the xauth key exported to the target will not
be destroyed in the target's xauth database until the source user
closes the last session.  The reference counts are kept in the file
~/.xauth/refcount/\fItargetname\fP/\fIkeyname\fP along with the
keys.  If the key for a display changes, the reference count will
drop to zero.
.SH EXAMPLES
While pam_xauth will create the .xauth directory the first time it is
run, if you want to set it up before it has ever run, you can do it
yourself like so:
.br
\f(CW$ mkdir ~/.xauth
.br
$ chmod 700 ~/.xauth\fR
.br
The rest of these examples assume that the ~/.xauth directory exists.

A user does not want to export any xauth keys, even to the
root account:
.br
\f(CW$ touch ~/.xauth/export\fR
.br

A system administrator often needs to su to his/her own account while
working from other people's accounts to fix their problems.  He/she
does not want his/her xauth database polluted with extra keys:
.br
\f(CW$ touch ~/.xauth/import\fR
.br

A user often helps his/her friends with their accounts and wants X to
work when he has used su to assume their accounts, no matter who the
user is:
.br
\f(CW$ echo '*' > ~/.xauth/export\fR

Frank trusts Bill and the root user, but no one else, with his xauth
keys:
.br
\f(CW$ echo 'root
.br
bill' > ~/.xauth/export\fR
.SH ARGUMENTS
.IP debug
Turns on debugging messages sent to syslog.  Most debugging messages are
sent as AUTHPRIV messages, which means that on most systems they will
show up in /var/log/secure.
.IP logpub
Makes none of the syslogged messages be sent AUTHPRIV.  This is dangerous
on a multi-user system but can make debugging much simpler because all
the logging is in one place.  Normally only used with the \fBdebug\fP
argument.
.IP warndays=\fIdays\fP
Obsolete, ignored.
.IP warnhours=\fIhours\fP
Obsolete, ignored.
.IP systemuser=\fInumericuserid\fP
Treat all users with ids less than or equal to \fInumericuserid\fP but
greater than 0 (that is; excepting the root user) as
system users, and do not forward keys to them or from them.  Set
systemuser=0 to disable this checking entirely.  The default is
499.
.IP xauthpath=\fI/usr/X11R6/bin/xauth\fP
Specify the path the xauth program (the default is /usr/X11R6/bin/xauth).
.SH "IMPLEMENTATION DETAILS"
pam_xauth will work \fIonly\fP if it is used from a setuid application
in which the getuid() call returns the id of the user running the
application (alternatively, it will work if the getlogin() call works,
but that is out of control of many of the target applications, like su),
and for which PAM can supply the name of the account that
the user is attempting to assume.  The typical application of this
type is su.  The application must call both pam_open_session() and
pam_close_session() with the ruid set to the id of the calling user
and the euid set to root, and must have provided as the PAM_USER item
the name of the target user.

pam_xauth calls xauth as the source user to find the canonical key
name and extract the key for
$DISPLAY, then calls xauth as the target user to merge the key
into the target user's database and later remove the key.

pam_xauth cannot be told not to remove the keys when the session
is closed.  If, for example, you wish for the root account to
collect all possible keys and never remove them from the database,
you should disable pam_xauth key import by creating an empty
~root/.xauth/import file, then set up root's .profile (or whatever)
to do something like this:
.nf
\f(CWif [ -n "$DISPLAY" -a -n "$LOGNAME" -a "$LOGNAME" != root ]
then
    su -c $LOGNAME xauth -iq nextract - "$DISPLAY" | \e
        xauth nmerge -
fi\fR
.fi
.SH FILES
~/.xauth/export
.br
~/.xauth/import
.br
~/.xauth/refcount/ and all file and subdirectories contained therein
.SH "SEE ALSO"
\fI/usr/doc/pam*/html/index.html\fP
.SH BUGS
Let's hope not, but if you find any, please report them via the "Bug Track"
link at http://developer.redhat.com/
.SH AUTHOR
Michael K. Johnson <johnsonm@redhat.com>


NEWS
2004-06-14: revise to build most modules outside of the Linux-PAM build
	directory.  The lone exception is pam_stack, which requires a private
	header which wouldn't be available otherwise.


pam_xauth/pam_xauth.c
/*
 * Copyright 2001 Red Hat, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, and the entire permission notice in its entirety,
 *    including the disclaimer of warranties.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote
 *    products derived from this software without specific prior
 *    written permission.
 *
 * ALTERNATIVELY, this product may be distributed under the terms of
 * the GNU Public License, in which case the provisions of the GPL are
 * required INSTEAD OF the above restrictions.  (This clause is
 * necessary due to a potential bad interaction between the GPL and
 * the restrictions contained in a BSD-style copyright.)
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ident "$Id$"

#include "../../_pam_aconf.h"
#include <sys/types.h>
#include <sys/wait.h>
#include <errno.h>
#include <limits.h>
#include <netdb.h>
#include <pwd.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#define PAM_SM_SESSION
#include "../../libpam/include/security/pam_modules.h"

#define PAM_GETPWNAM_R
#define PAM_GETPWUID_R
#include "../../libpam/include/security/_pam_macros.h"

#define DATANAME "pam_xauth_cookie_file"
#define XAUTHBIN "/usr/X11R6/bin/xauth"
#define XAUTHENV "XAUTHORITY"
#define HOMEENV  "HOME"
#define XAUTHDEF ".Xauthority"
#define XAUTHTMP ".xauthXXXXXX"

/* Run a given command (with a NULL-terminated argument list), feeding it the
 * given input on stdin, and storing any output it generates. */
static int
run_coprocess(const char *input, char **output,
	      uid_t uid, gid_t gid, const char *command, ...)
{
	int ipipe[2], opipe[2], i;
	char buf[LINE_MAX];
	pid_t child;
	char *buffer = NULL;
	size_t buffer_size = 0;
	va_list ap;

	*output = NULL;

	/* Create stdio pipery. */
	if (pipe(ipipe) == -1) {
		return -1;
	}
	if (pipe(opipe) == -1) {
		close(ipipe[0]);
		close(ipipe[1]);
		return -1;
	}

	/* Fork off a child. */
	child = fork();
	if (child == -1) {
		close(ipipe[0]);
		close(ipipe[1]);
		close(opipe[0]);
		close(opipe[1]);
		return -1;
	}

	if (child == 0) {
		/* We're the child. */
		char *args[10];
		const char *tmp;
		/* Drop privileges. */
		setgid(gid);
		setgroups(0, NULL);
		setuid(uid);
		/* Initialize the argument list. */
		memset(&args, 0, sizeof(args));
		/* Set the pipe descriptors up as stdin and stdout, and close
		 * everything else, including the original values for the
		 * descriptors. */
		dup2(ipipe[0], STDIN_FILENO);
		dup2(opipe[1], STDOUT_FILENO);
		for(i = 0; i < sysconf(_SC_OPEN_MAX); i++) {
			if ((i != STDIN_FILENO) && (i != STDOUT_FILENO)) {
				close(i);
			}
		}
		/* Convert the varargs list into a regular array of strings. */
		va_start(ap, command);
		args[0] = strdup(command);
		for(i = 1; i < ((sizeof(args) / sizeof(args[0])) - 1); i++) {
			tmp = va_arg(ap, const char*);
			if (tmp == NULL) {
				break;
			}
			args[i] = strdup(tmp);
		}
		/* Run the command. */
		execvp(command, args);
		/* Never reached. */
		exit(1);
	}

	/* We're the parent, so close the other ends of the pipes. */
	close(ipipe[0]);
	close(opipe[1]);
	/* Send input to the process (if we have any), then send an EOF. */
	if (input) {
		write(ipipe[1], input, strlen(input));
	}
	close(ipipe[1]);

	/* Read data output until we run out of stuff to read. */
	i = read(opipe[0], buf, sizeof(buf));
	while((i != 0) && (i != -1)) {
		char *tmp;
		/* Resize the buffer to hold the data. */
		tmp = realloc(buffer, buffer_size + i + 1);
		if (tmp == NULL) {
			/* Uh-oh, bail. */
			if (buffer != NULL) {
				free(buffer);
			}
			close(opipe[0]);
			waitpid(child, NULL, 0);
			return -1;
		}
		/* Save the new buffer location, copy the newly-read data into
		 * the buffer, and make sure the result will be
		 * nul-terminated. */
		buffer = tmp;
		memcpy(buffer + buffer_size, buf, i);
		buffer[buffer_size + i] = '\0';
		buffer_size += i;
		/* Try to read again. */
		i = read(opipe[0], buf, sizeof(buf));
	}
	/* No more data.  Clean up and return data. */
	close(opipe[0]);
	*output = buffer;
	waitpid(child, NULL, 0);
	return 0;
}

/* Free a data item. */
static void
cleanup(pam_handle_t *pamh, void *data, int err)
{
	free(data);
}

int
pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	char xauthpath[] = XAUTHBIN;
	char *cookiefile = NULL, *xauthority = NULL,
	     *cookie = NULL, *display = NULL, *thome = NULL, *tmp;
	const char *user, *xauth = xauthpath;
	struct passwd passwd, *pwd;
	size_t buflen;
	int fd, i, debug = 0;
	uid_t systemuser = 499;

	/* Parse arguments.  We don't understand many, so no sense in breaking
	 * this into a separate function. */
	for(i = 0; i < argc; i++) {
		if (strcmp(argv[i], "debug") == 0) {
			debug = 1;
			continue;
		}
		if (strncmp(argv[i], "xauthpath=", 10) == 0) {
			xauth = argv[i] + 10;
			continue;
		}
		if (strncmp(argv[i], "systemuser=", 11) == 0) {
			long l = strtol(argv[i] + 11, &tmp, 10);
			if ((strlen(argv[i] + 11) > 0) && (*tmp == '\0')) {
				systemuser = l;
			} else {
				syslog(LOG_WARNING, "pam_xauth: invalid value "
				       "for systemuser (`%s')", argv[i] + 11);
			}
			continue;
		}
		syslog(LOG_WARNING, "pam_xauth: unrecognized option `%s'",
		       argv[i]);
	}

	/* If DISPLAY isn't set, we don't really care, now do we? */
	if ((display = getenv("DISPLAY")) == NULL) {
		if (debug) {
			syslog(LOG_DEBUG, "pam_xauth: user has no DISPLAY,"
			       " doing nothing");
		}
		return PAM_SUCCESS;
	}

	/* Read the target user's name. */
	if (pam_get_item(pamh, PAM_USER, (const void**)&user) != PAM_SUCCESS) {
		syslog(LOG_ERR, "pam_xauth: error determining target "
		       "user's name");
		return PAM_SESSION_ERR;
	}

	/* Get the target user's UID and primary GID, which we'll need to set
	 * on the xauthority file we create later on. */
	if (_pam_getpwnam_r(user, &passwd, &tmp, &buflen, &pwd) != 0) {
		syslog(LOG_ERR, "pam_xauth: error determining target "
		       "user's UID");
		return PAM_SESSION_ERR;
	}

	/* We only care about the target user's IDs, so we can free the
	 * string data after making a copy of the user's home directory. */
	thome = strdup(passwd.pw_dir);
	if (tmp) {
		free(tmp);
	}

	/* If the UID is a system account (and not the superuser), forget
	 * about forwarding keys. */
	if ((passwd.pw_uid != 0) && (passwd.pw_uid <= systemuser)) {
		free(thome);
		return PAM_SUCCESS;
	}

	/* Figure out where the source user's .Xauthority file is. */
	if (getenv(XAUTHENV) != NULL) {
		cookiefile = strdup(getenv(XAUTHENV));
	} else {
		char *t, *homedir = NULL;
		size_t t_len;
		struct passwd tpasswd, *tpwd;
		if (_pam_getpwuid_r(getuid(), &tpasswd, &t, &t_len,
				   &tpwd) == 0) {
			homedir = strdup(tpasswd.pw_dir);
			free(t);
		} else {
			free(thome);
			return PAM_SESSION_ERR;
		}
		cookiefile = malloc(strlen(homedir) + 1 + strlen(XAUTHDEF) + 1);
		if (cookiefile == NULL) {
			free(thome);
			free(homedir);
			return PAM_SESSION_ERR;
		}
		strcpy(cookiefile, homedir);
		strcat(cookiefile, "/");
		strcat(cookiefile, XAUTHDEF);
		free(homedir);
	}
	if (debug) {
		syslog(LOG_DEBUG, "pam_xauth: reading keys from `%s'",
		       cookiefile);
	}

	/* Read the user's .Xauthority file.  Because the current UID is
	 * the original user's UID, this will only fail if something has
	 * gone wrong, or we have no cookies. */
	if (run_coprocess(NULL, &cookie,
			  getuid(), getgid(),
			  xauth, "-f", cookiefile, "nlist", display,
			  NULL) == 0) {
		/* Check that we got a cookie.  If not, we get creative. */
		if (((cookie == NULL) || (strlen(cookie) == 0)) &&
		    ((strncmp(display, "localhost:", 10) == 0) ||
		     (strncmp(display, "localhost/unix:", 15) == 0))) {
			char *t, *screen;
			size_t tlen, slen;
			/* Free the useless cookie string. */
			if (cookie != NULL) {
				free(cookie);
				cookie = NULL;
			}
			/* Allocate enough space to hold an adjusted name. */
			tlen = strlen(display) + LINE_MAX + 1;
			t = malloc(tlen);
			if (t != NULL) {
				memset(t, 0, tlen);
				if (gethostname(t, tlen - 1) != -1) {
					/* Append the protocol and then the
					 * screen number. */
					if (strlen(t) < tlen - 6) {
						strcat(t, "/unix:");
					}
					screen = strchr(display, ':');
					if (screen != NULL) {
						screen++;
						slen = strlen(screen);
						if (strlen(t) + slen < tlen) {
							strcat(t, screen);
						}
					}
					if (debug) {
						syslog(LOG_DEBUG, "pam_xauth: "
						       "no key for `%s', trying"
						       " `%s'", display, t);
					}
					/* Read the cookie for this display. */
					run_coprocess(NULL, &cookie,
						      getuid(), getgid(),
						      xauth, "-f", cookiefile,
						      "nlist", t, NULL);
				}
				free(t);
			}
		}

		/* Check that we got a cookie, this time for real. */
		if ((cookie == NULL) || (strlen(cookie) == 0)) {
			if (debug) {
				syslog(LOG_DEBUG, "pam_xauth: no key");
			}
			free(thome);
			free(cookiefile);
			return PAM_SESSION_ERR;
		}

		/* Generate the environment variable
		 * "XAUTHORITY=<homedir>/filename". */
		xauthority = malloc(strlen(XAUTHENV) + strlen(thome) +
				    strlen(XAUTHTMP) + 3);
		if (xauthority == NULL) {
			if (debug) {
				syslog(LOG_DEBUG, "pam_xauth: no free memory");
			}
			free(thome);
			free(cookiefile);
			free(cookie);
			return PAM_SESSION_ERR;
		}
		strcpy(xauthority, XAUTHENV);
		strcat(xauthority, "=");
		strcat(xauthority, thome);
		strcat(xauthority, "/");
		strcat(xauthority, XAUTHTMP);

		/* Generate a new file to hold the data. */
		fd = mkstemp(xauthority + strlen(XAUTHENV) + 1);
		if (fd == -1) {
			syslog(LOG_ERR, "pam_xauth: error creating "
			       "temporary file `%s': %s",
			       xauthority + strlen(XAUTHENV) + 1,
			       strerror(errno));
			free(thome);
			free(cookiefile);
			free(cookie);
			free(xauthority);
			return PAM_SESSION_ERR;
		}
		/* Set permissions on the new file and dispose of the
		 * descriptor. */
		fchown(fd, passwd.pw_uid, passwd.pw_gid);
		close(fd);

		/* Get a copy of the filename to save as a data item for
		 * removal at session-close time. */
		free(cookiefile);
		cookiefile = strdup(xauthority + strlen(XAUTHENV) + 1);

		/* Save the filename. */
		if (pam_set_data(pamh, DATANAME, cookiefile, cleanup) != PAM_SUCCESS) {
			syslog(LOG_ERR, "pam_xauth: error saving name of "
			       "temporary file `%s'", cookiefile);
			unlink(cookiefile);
			free(thome);
			free(xauthority);
			free(cookiefile);
			free(cookie);
			return PAM_SESSION_ERR;
		}

		/* Unset any old XAUTHORITY variable in the environment. */
		if (getenv(XAUTHENV)) {
			unsetenv(XAUTHENV);
		}

		/* Set the new variable in the environment. */
		pam_putenv(pamh, xauthority);
		putenv(xauthority); /* The environment owns this string now. */

		/* Merge the cookie we read before into the new file. */
		if (debug) {
			syslog(LOG_DEBUG, "pam_xauth: writing key `%s' to "
			       "temporary file `%s'", cookie, cookiefile);
		}
		run_coprocess(cookie, &tmp,
			      passwd.pw_uid, passwd.pw_gid,
			      xauth, "-f", cookiefile, "nmerge", "-", NULL);

		/* We don't need to keep a copy of these around any more. */
		free(cookie);
		free(thome);
	}
	return PAM_SUCCESS;
}

int
pam_sm_close_session(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	void *cookiefile;
	int i, debug = 0;

	/* Parse arguments.  We don't understand many, so no sense in breaking
	 * this into a separate function. */
	for(i = 0; i < argc; i++) {
		if (strcmp(argv[i], "debug") == 0) {
			debug = 1;
			continue;
		}
		if (strncmp(argv[i], "xauthpath=", 10) == 0) {
			continue;
		}
		if (strncmp(argv[i], "systemuser=", 11) == 0) {
			continue;
		}
		syslog(LOG_WARNING, "pam_xauth: unrecognized option `%s'",
		       argv[i]);
	}

	/* Try to retrieve the name of a file we created when the session was
	 * opened. */
	if (pam_get_data(pamh, DATANAME, (const void**) &cookiefile) == PAM_SUCCESS) {
		/* We'll only try to remove the file once. */
		if (strlen((char*)cookiefile) > 0) {
			if (debug) {
				syslog(LOG_DEBUG, "pam_xauth: removing `%s'",
				       (char*)cookiefile);
			}
			unlink((char*)cookiefile);
			*((char*)cookiefile) = '\0';
		}
	}
	return PAM_SUCCESS;
}
